File: JavaScript/src/Internals/Functions/Operator/Function_AND_N.js
import { Function_N } from '../Function_N.js';
import { Operand } from '../../../Operand.js';

class Function_AND_N extends Function_N {
  get Name() {
    return 'AndN';
  }

  constructor(z) {
    super(z);
  }

  evaluate(work, tempParameter) {
    let b = true;
    for (let i = 0; i < this.z.length; i++) {
      let a = this.getBoolean(work, tempParameter, i);
      if (a.IsErrorOrNone) { return a; }
      if (a.BooleanValue === false) {
        b = false;
        // 非严格模式短路,严格模式继续求值以检查后续错误
        if (work.UseStrictMode === false) { break; }
      }
    }
    return b ? Operand.True : Operand.False;
  }
}

export { Function_AND_N };

File: java/src/main/java/toolgood/algorithm/internals/functions/operator/Function_AND_N.java
package toolgood.algorithm.internals.functions.operator;

import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.Operand;
import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.internals.functions.Function_N;

public class Function_AND_N extends Function_N {
    public Function_AND_N(FunctionBase[] funcs) {
        super(funcs);
    }

    @Override
    public Operand Evaluate(AlgorithmEngine work, java.util.function.BiFunction<AlgorithmEngine, String, Operand> tempParameter) {
        int index = 1;
        boolean b = true;
        for (FunctionBase item : funcs) {
            Operand a = item.Evaluate(work, tempParameter);
            if (a.IsNotBoolean()) {
                a = a.ToBoolean("Function '{0}' parameter {1} is error!", "AND", index++);
                if (a.IsError()) {
                    return a;
                }
            }
            if (a.BooleanValue() == false) {
                b = false;
            }
        }
        return b ? Operand.TRUE : Operand.FALSE;
    }

    @Override
    public void toString(StringBuilder stringBuilder, boolean addBrackets) {
        AddFunction(stringBuilder, "AND");
    }
}
